File: leetecode/CombinationSum.cpp
#include <iostream>
#include <vector>
#include <string>
#include <unordered_set>
#include <algorithm>
using namespace std;
int hashMe(vector<int> v)
{
    sort(v.begin(),v.end());
    std::hash<int> hasher;
    size_t seed = 0;
    for (int i : v)
    {
        seed ^= hasher(i) + 0x9e3779b9 + (seed << 6) + (seed >> 2);
    }
    return seed;
}
void rec(int target, vector<int> res, vector<vector<int>> &allRes, vector<int> &candidates, unordered_set<int> &vc)
{
    if (target < 0)
    {
        return;
    }
    if (target == 0)
    {
        int hash = hashMe(res);
        if (vc.count(hash))
        {
            return;
        }
        allRes.push_back(res);
        vc.insert(hash);
        return;
    }
    for (auto a : candidates)
    {
        res.push_back(a);
        rec(target - a, res, allRes, candidates, vc);
        res.pop_back();
    }
}
vector<vector<int>> combinationSum(vector<int> &candidates, int target)
{
    vector<vector<int>> allRes = {};
    unordered_set<int> vc;
        rec(target, {}, allRes, candidates, vc);
    return allRes;
}
int main()
{

    return 0;
}

File: leetecode/CombinationSum.java


import java.util.ArrayList;
import java.util.List;

public class CombinationSum {
    List<List<Integer>> allRes;
    private void rec(int[] candidates,int target,int cur,ArrayList<Integer> path,int start){
        if(cur>target) return;
        if(cur==target){
            allRes.add((List<Integer>) path.clone());
            return;
        }
        if(path==null) path = new ArrayList<>();
        for(int i=start;i<candidates.length;i++) {
            path.add(candidates[i]);
            rec(candidates,target,cur+candidates[i],path, i+1);
            path.remove(path.size()-1);
        }
    }
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        allRes = new ArrayList<>();
        rec(candidates,target,0,null,0);
        return allRes;
    }
}
